File: src/mesh3d.py
"""Mesh representation for the cloth simulation."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable, List, Tuple

import numpy as np


@dataclass
class Mesh3D:
    """Simple 3D mesh tailored for cloth simulations.

    The mesh stores vertex positions and a list of edges.  Each vertex has a mask
    that defines which translational degrees of freedom (x, y, z) are free.  This
    enables the simulation to constrain nodes individually along any axis.
    """

    positions: np.ndarray
    edges: List[Tuple[int, int]]
    free_dof: np.ndarray
    faces: List[Tuple[int, int, int]] = field(default_factory=list)
    initial_positions: np.ndarray = field(init=False)
    base_free_dof: np.ndarray = field(init=False)

    def __post_init__(self) -> None:
        self.positions = np.asarray(self.positions, dtype=np.float64)
        if self.positions.ndim != 2 or self.positions.shape[1] != 3:
            raise ValueError("positions must be an (N, 3) array")

        self.initial_positions = self.positions.copy()

        if len(self.edges) == 0:
            raise ValueError("At least one edge is required for the simulation")

        self.free_dof = np.asarray(self.free_dof, dtype=bool)
        if self.free_dof.shape != self.positions.shape:
            raise ValueError("free_dof must match the shape of positions")
        self.base_free_dof = self.free_dof.copy()

        # Faces are optional for simulation but required for shaded rendering.
        self.faces = [tuple(face) for face in self.faces]

    @property
    def n_vertices(self) -> int:
        return self.positions.shape[0]

    def fix_vertex(
        self,
        index: int,
        axes: Iterable[int] = (0, 1, 2),
        *,
        permanent: bool = True,
    ) -> None:
        """Constrains the vertex along the provided axes.

        Parameters
        ----------
        index:
            Index of the vertex to lock.
        axes:
            Iterable with the axis indices (0=x, 1=y, 2=z) to constrain.
        permanent:
            When ``True`` (default) the constraint is stored as part of the
            mesh's baseline state so that :meth:`reset` keeps the vertex fixed.
        """

        axes = tuple(axes)
        for axis in axes:
            self.free_dof[index, axis] = False
            self.positions[index, axis] = self.initial_positions[index, axis]
            if permanent:
                self.base_free_dof[index, axis] = False

    def release_vertex(
        self,
        index: int,
        axes: Iterable[int] = (0, 1, 2),
        *,
        permanent: bool = True,
    ) -> None:
        axes = tuple(axes)
        for axis in axes:
            self.free_dof[index, axis] = True
            if permanent:
                self.base_free_dof[index, axis] = True

    def axis_is_free(self, index: int, axis: int) -> bool:
        return bool(self.free_dof[index, axis])

    def clone(self) -> "Mesh3D":
        clone = Mesh3D(
            self.positions.copy(),
            list(self.edges),
            self.free_dof.copy(),
            list(self.faces),
        )
        clone.initial_positions = self.initial_positions.copy()
        clone.base_free_dof = self.base_free_dof.copy()
        return clone

    @staticmethod
    def build_square(size: float, subdivisions: int) -> "Mesh3D":
        """Builds a square cloth grid lying on the XY plane.

        The square is centered at the origin and subdivided into a regular grid.
        The four original corners are fully constrained (no movement).
        """

        if subdivisions < 1:
            raise ValueError("subdivisions must be >= 1")

        grid_n = subdivisions + 1
        step = size / subdivisions

        positions: List[Tuple[float, float, float]] = []
        free_dof: List[Tuple[bool, bool, bool]] = []
        edges: List[Tuple[int, int]] = []
        faces: List[Tuple[int, int, int]] = []

        # create vertices
        half = size / 2.0
        for iy in range(grid_n):
            for ix in range(grid_n):
                x = -half + ix * step
                y = -half + iy * step
                positions.append((x, y, 0.0))
                free_dof.append([True, True, True])

        # create structural edges (horizontal and vertical)
        def idx(ix: int, iy: int) -> int:
            return iy * grid_n + ix

        for iy in range(grid_n):
            for ix in range(grid_n):
                if ix + 1 < grid_n:
                    edges.append((idx(ix, iy), idx(ix + 1, iy)))
                if iy + 1 < grid_n:
                    edges.append((idx(ix, iy), idx(ix, iy + 1)))
                # add shear (diagonal) springs for stability
                if ix + 1 < grid_n and iy + 1 < grid_n:
                    edges.append((idx(ix, iy), idx(ix + 1, iy + 1)))
                if ix > 0 and iy + 1 < grid_n:
                    edges.append((idx(ix, iy), idx(ix - 1, iy + 1)))

                if ix + 1 < grid_n and iy + 1 < grid_n:
                    v0 = idx(ix, iy)
                    v1 = idx(ix + 1, iy)
                    v2 = idx(ix, iy + 1)
                    v3 = idx(ix + 1, iy + 1)
                    faces.append((v0, v1, v2))
                    faces.append((v2, v1, v3))

        mesh = Mesh3D(
            np.array(positions),
            edges,
            np.array(free_dof, dtype=bool),
            faces,
        )

        # Fix the four corners of the original square completely.
        corners = [idx(0, 0), idx(grid_n - 1, 0), idx(0, grid_n - 1), idx(grid_n - 1, grid_n - 1)]
        for corner in corners:
            mesh.fix_vertex(corner, axes=(0, 1, 2))

        return mesh

    def reset(self) -> None:
        self.positions[:] = self.initial_positions
        self.free_dof[:, :] = self.base_free_dof

    def compute_vertex_normals(self) -> np.ndarray:
        """Compute per-vertex normals from the available triangle faces."""

        normals = np.zeros_like(self.positions)
        for face in self.faces:
            i0, i1, i2 = face
            p0, p1, p2 = self.positions[[i0, i1, i2]]
            edge1 = p1 - p0
            edge2 = p2 - p0
            face_normal = np.cross(edge1, edge2)
            norm = np.linalg.norm(face_normal)
            if norm == 0:
                continue
            face_normal /= norm
            normals[i0] += face_normal
            normals[i1] += face_normal
            normals[i2] += face_normal

        norms = np.linalg.norm(normals, axis=1)
        nonzero = norms > 0
        normals[nonzero] /= norms[nonzero][:, None]
        # Default normal for isolated vertices.
        normals[~nonzero] = np.array([0.0, 0.0, 1.0])
        return normals

File: cpp/Mesh3D.cpp
#include "Mesh3D.h"

#include <stdexcept>

Mesh3D::Mesh3D(
    std::vector<Vec3> positions_in,
    std::vector<std::pair<int, int>> edges_in,
    std::vector<std::array<bool, 3>> free_dof_in,
    std::vector<std::array<int, 3>> faces_in
) :
    positions(std::move(positions_in)),
    edges(std::move(edges_in)),
    free_dof(std::move(free_dof_in)),
    faces(std::move(faces_in)),
    initial_positions_(positions),
    base_free_dof_(free_dof)
{
    if (positions.empty()) {
        throw std::invalid_argument("positions cannot be empty");
    }
    if (edges.empty()) {
        throw std::invalid_argument("edges cannot be empty");
    }
    if (free_dof.size() != positions.size()) {
        throw std::invalid_argument("free_dof must match positions length");
    }
}

Mesh3D Mesh3D::BuildSquare(double size, int subdivisions) {
    if (subdivisions < 1) {
        throw std::invalid_argument("subdivisions must be >= 1");
    }

    int grid_n = subdivisions + 1;
    double step = size / static_cast<double>(subdivisions);

    std::vector<Vec3> positions;
    std::vector<std::array<bool, 3>> free_dof;
    std::vector<std::pair<int, int>> edges;
    std::vector<std::array<int, 3>> faces;

    positions.reserve(grid_n * grid_n);
    free_dof.reserve(grid_n * grid_n);

    double half = size / 2.0;
    for (int iy = 0; iy < grid_n; ++iy) {
        for (int ix = 0; ix < grid_n; ++ix) {
            double x = -half + ix * step;
            double y = -half + iy * step;
            positions.emplace_back(x, y, 0.0);
            free_dof.push_back({true, true, true});
        }
    }

    auto idx = [grid_n](int ix, int iy) {
        return iy * grid_n + ix;
    };

    for (int iy = 0; iy < grid_n; ++iy) {
        for (int ix = 0; ix < grid_n; ++ix) {
            if (ix + 1 < grid_n) {
                edges.emplace_back(idx(ix, iy), idx(ix + 1, iy));
            }
            if (iy + 1 < grid_n) {
                edges.emplace_back(idx(ix, iy), idx(ix, iy + 1));
            }
            if (ix + 1 < grid_n && iy + 1 < grid_n) {
                edges.emplace_back(idx(ix, iy), idx(ix + 1, iy + 1));
            }
            if (ix > 0 && iy + 1 < grid_n) {
                edges.emplace_back(idx(ix, iy), idx(ix - 1, iy + 1));
            }

            if (ix + 1 < grid_n && iy + 1 < grid_n) {
                int v0 = idx(ix, iy);
                int v1 = idx(ix + 1, iy);
                int v2 = idx(ix, iy + 1);
                int v3 = idx(ix + 1, iy + 1);
                faces.push_back({v0, v1, v2});
                faces.push_back({v2, v1, v3});
            }
        }
    }

    Mesh3D mesh(std::move(positions), std::move(edges), std::move(free_dof), std::move(faces));

    std::vector<int> corners = {
        idx(0, 0),
        idx(grid_n - 1, 0),
        idx(0, grid_n - 1),
        idx(grid_n - 1, grid_n - 1)
    };
    for (int corner : corners) {
        mesh.FixVertex(corner);
    }

    return mesh;
}

void Mesh3D::FixVertex(int index, const std::vector<int>& axes, bool permanent) {
    for (int axis : axes) {
        if (axis < 0 || axis >= 3) {
            continue;
        }
        free_dof[index][axis] = false;
        positions[index].x = initial_positions_[index].x;
        positions[index].y = initial_positions_[index].y;
        positions[index].z = initial_positions_[index].z;
        if (permanent) {
            base_free_dof_[index][axis] = false;
        }
    }
}

void Mesh3D::ReleaseVertex(int index, const std::vector<int>& axes, bool permanent) {
    for (int axis : axes) {
        if (axis < 0 || axis >= 3) {
            continue;
        }
        free_dof[index][axis] = true;
        if (permanent) {
            base_free_dof_[index][axis] = true;
        }
    }
}

bool Mesh3D::AxisIsFree(int index, int axis) const {
    if (axis < 0 || axis >= 3) {
        return false;
    }
    return free_dof[index][axis];
}

void Mesh3D::Reset() {
    positions = initial_positions_;
    free_dof = base_free_dof_;
}

void Mesh3D::Translate(const Vec3& delta, bool update_initial) {
    for (auto& pos : positions) {
        pos += delta;
    }
    if (update_initial) {
        for (auto& init : initial_positions_) {
            init += delta;
        }
    }
}

std::vector<Vec3> Mesh3D::ComputeVertexNormals() const {
    std::vector<Vec3> normals(positions.size(), Vec3(0.0, 0.0, 0.0));
    for (const auto& face : faces) {
        int i0 = face[0];
        int i1 = face[1];
        int i2 = face[2];
        const Vec3& p0 = positions[i0];
        const Vec3& p1 = positions[i1];
        const Vec3& p2 = positions[i2];
        Vec3 edge1 = p1 - p0;
        Vec3 edge2 = p2 - p0;
        Vec3 face_normal = Cross(edge1, edge2);
        double norm = face_normal.Norm();
        if (norm == 0.0) {
            continue;
        }
        face_normal /= norm;
        normals[i0] += face_normal;
        normals[i1] += face_normal;
        normals[i2] += face_normal;
    }

    for (auto& normal : normals) {
        double len = normal.Norm();
        if (len == 0.0) {
            normal = Vec3(0.0, 0.0, 1.0);
        } else {
            normal /= len;
        }
    }
    return normals;
}
